scripts/cybershake/cybershake_progress.py

```diff
@@ -0,0 +1,364 @@
+    column_t.append(("total", EST_CORE_HOURS_COL))
+    column_t.append(("total", ACT_CORE_HOURS_COL))
+
+    progress_df = pd.DataFrame(
+        index=grouped_df.index.values, columns=pd.MultiIndex.from_tuples(column_t)
+    )
+
+    # Populate progress dataframe with estimation data
+    total = np.nan
+    for proc_type in PROCESS_TYPES:
+        values = grouped_df[proc_type, const.MetadataField.core_hours.value]
+        progress_df[proc_type, EST_CORE_HOURS_COL] = values
+        total = values if total is np.nan else total + values
+    progress_df["total", EST_CORE_HOURS_COL] = total
+    progress_df["total", R_COUNT_COL] = r_counts
+    # Get actual core hours for all faults, assumes faults are run
+    # in alphabetical order
+    chours_df = get_chours_used(faults)
+
+    # Add actual data to progress df
+    total = np.nan
+    for proc_type in PROCESS_TYPES:
+        values = chours_df[proc_type]
+        progress_df[proc_type, ACT_CORE_HOURS_COL] = values
+        total = values if total is np.nan else total + values
+    progress_df["total", ACT_CORE_HOURS_COL] = total
+
+    return progress_df
+
+
+def load_progress_df(file: str):
+    return pd.read_csv(file, index_col=[0], header=[0, 1])
+
+
+def print_progress(
+    progress_df: pd.DataFrame, outfile: str, db_file: str, cur_fault_ix: int = None
+):
+    """Prints the progress dataframe in a nice format.
+
+    If cur_fault_ix is specified then all uncompleted faults are printed and
+    the latest completed fault with the previous and next five faults.
+
+    Otherwise all faults are printed.
+    """
+
+    f = open(outfile, "w")
+
+    progress_df.to_csv("progress_df.csv")
+
+    def get_usage_str(fault: str, proc_type: str):
+        return "{:.2f}/{:.2f}".format(
+            progress_df.loc[fault, (proc_type, ACT_CORE_HOURS_COL)],
+            progress_df.loc[fault, (proc_type, EST_CORE_HOURS_COL)],
+        )
+
+    to_print_mask = np.ones(progress_df.shape[0], dtype=bool)
+
+    template_fmt = "{:<15}{:<22}{:<7}{:<3}{:<17}{:<7}{:<3}{:<17}{:<7}{:<3}{:<22}"
+    f.write(
+        template_fmt.format(
+            "Fault name",
+            "LF time",
+            "rels",
+            "C",
+            "HF time",
+            "rels",
+            "C",
+            "BB time",
+            "rels",
+            "C",
+            "Total",
+        )
+    )
+    f.write("\n")
+
+    est_completed_all = {}
+    for fault in progress_df.index.values[to_print_mask]:
+        est_completed_all[fault] = {}
+        num_all_rels = int(
+            progress_df.loc[fault, ("total", R_COUNT_COL)]
+        )  # not sure why this is np.float64
+        num_complete_rels = []
+        usage_str_list = []
+        is_complete = {}
+        for proc_type in PROCESS_TYPES:
+            usage_str_list.append(get_usage_str(fault, proc_type))
+            num_complete = get_completed_r_count(db_file, proc_type, fault)
+            num_complete_rels.append(f"{num_complete}/{num_all_rels}")
```

Comment:
For associative data like this I generally find dictionaries more logical rather than using implicit ordering.

Reply:
Sure